// @solidstate/hardhat-solidstate-utils/src/lib/print.ts
import type { PackageData } from '../types.js';
import banner from './banner.js';
import { createTable } from './table.js';
import chalk from 'chalk';
import semver from 'semver';

export const printSolidstateFunding = () => {
  console.log(banner);

  console.log('This is a Solidstate Hardhat project.');

  // TODO: print funding links
};

export const printSolidstatePluginStatus = (
  solidstatePlugins: PackageData[] = [],
) => {
  solidstatePlugins.sort((a, b) => (a.name > b.name ? 1 : -1));

  const table = createTable();

  table.push([
    { content: chalk.bold('Package') },
    { content: chalk.bold('Version') },
    { content: chalk.bold('NPM Link') },
    { content: chalk.bold('Status') },
  ]);

  for (const solidstatePackage of solidstatePlugins) {
    const { name, latestVersion, installedVersion, loaded } = solidstatePackage;

    table.push([
      {
        content:
          name.replace('@solidstate/', chalk.gray('@solidstate/')) +
          (latestVersion ? chalk.gray(`@${latestVersion}`) : ''),
      },
      { content: installedVersion ?? '' },
      {
        content: '☍',
        hAlign: 'center',
        href: latestVersion ? `https://npmjs.com/package/${name}` : undefined,
      },
      {
        content: installedVersion
          ? loaded
            ? !latestVersion || semver.gte(installedVersion, latestVersion)
              ? chalk.green('●')
              : chalk.yellow('●')
            : chalk.red('●')
          : chalk.gray('●'),
        hAlign: 'center',
      },
    ]);
  }

  console.log(table.toString());
};
